src/Yoktorm/DbContext.cs

```diff
@@ -5,44 +5,42 @@
 using System.Text;
 using System.Threading.Tasks;
 using Yoktorm.ObjectState;
+using Dapper;
 
 namespace Yoktorm
 {
     /// <summary>
     /// Yoktorm database context. This context manage all statement executions, connections, model states and is the public interface of the orm
     /// </summary>
     /// <typeparam name="TConnection">ADO.Net Connection class of the ado.net provider that should be used</typeparam>
-    public abstract class DbContext<TConnection> : IDbContext where TConnection : IDbConnection, IDisposable
+    public abstract class DbContext<TConnection> : IDbContext, IDisposable, IDynamicDbContext where TConnection : IDbConnection
     {
         #region Fields
-        private IConnectionHandler<TConnection> connectionHandler;
-        private IProvider provider;
         private ObjectState.IObjectStateManager objectStateManager;
         private IDbConnection connection;
         private string connectionString;
-
-        // Database which will be used for caching information
-        private string database;
+        private IDatabase database;
+        private object _lock = new object();
         #endregion
 
         #region Constructor
         /// <summary>
         /// Initialize new database context
         /// </summary>
         /// <param name="provider">Provider instance</param>
-        /// <param name="connectionHandler">Connection handler</param>
         /// <param name="connectionString">Connection string</param>
         /// <param name="useObjectStateManager">Defines whether to use the object state manager or not</param>
-        internal DbContext(IDatabase database, IProvider provider, IConnectionHandler<TConnection> connectionHandler, string connectionString, bool useObjectStateManager)
+        protected DbContext(IDatabase database, string connectionString, bool useObjectStateManager)
         {
-            this.provider = provider;
-            this.connectionHandler = connectionHandler;
+            this.database = database;
             this.connectionString = connectionString;
 
             if (useObjectStateManager)
             {
                 this.objectStateManager = CreateObjectStateManager();
             }
+
+            InitializeDatabase();
```

Comment:
We need to check, whether it is a good idea that the first DbContext initializes the database. Also look here: https://github.com/simplic/yoktorm/pull/13/files#diff-f10fc4f64485ed1f84b72cadff7ffe7dR62 @yvz5